exporter/splunkhecexporter/client.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package splunkhecexporter
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.uber.org/zap"
+)
+
+// client sends the data to the splunk backend.
+type client struct {
+	config  *Config
+	url     *url.URL
+	client  *http.Client
+	logger  *zap.Logger
+	zippers sync.Pool
+	wg sync.WaitGroup
+}
+
+func (c *client) pushMetricsData(
+	ctx context.Context,
+	md consumerdata.MetricsData,
+) (droppedTimeSeries int, err error) {
+	c.wg.Add(1)
+	defer c.wg.Done()
+
+	splunkDataPoints, numDroppedTimeseries, err := metricDataToSplunk(c.logger, md, c.config)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	body, compressed, err := c.encodeBody(splunkDataPoints)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	req, err := http.NewRequest("POST", c.url.String(), body)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+	headers := buildHeaders(c.config)
+
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+
+	if compressed {
+		req.Header.Set("Content-Encoding", "gzip")
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), err
+	}
+
+	io.Copy(ioutil.Discard, resp.Body)
+	resp.Body.Close()
+
+	// Splunk accepts all 2XX codes.
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		err = fmt.Errorf(
+			"HTTP %d %q",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+		return exporterhelper.NumTimeSeries(md), err
+	}
+
+	return numDroppedTimeseries, nil
+}
+
+func buildHeaders(config *Config) map[string]string {
+	headers := map[string]string{
+		"Connection":   "keep-alive",
+		"Content-Type": "application/json",
+		"User-Agent":   "OpenTelemetry-Collector Splunk Exporter/v0.0.1",
+	}
+
+	if config.Token != "" {
+		headers["Authorization"] = "Splunk " + config.Token
+	}
+
+	return headers
+}
+
+func (c *client) encodeBody(dps []*splunkMetric) (bodyReader io.Reader, compressed bool, err error) {
```

Comment:
Looks like this and getReader are just helper functions, don't need state inside client so could just make them plain functions.

Reply:
there was some state reference, but easy enough to pass in as parameter. Refactored.